crates/solver/src/driver.rs

```diff
@@ -368,16 +368,43 @@ impl Driver {
         let run = self.next_run_id();
 
         // extra function so that we can add span information
-        self.single_auction(auction, run)
+        let stored_competition_info = self
+            .single_auction(auction, run)
             .instrument(tracing::info_span!("auction", id, run))
-            .await
+            .await?;
+        // Wait for the auction to notice that we submitted competition info and update the next id.
+        // Otherwise there is a chance that we use the same id again.
+        if stored_competition_info {
+            let wait_for_id_to_change = async {
+                tracing::debug!("waiting for auction competition id to change");
+                loop {
+                    tokio::time::sleep(Duration::from_secs(1)).await;
+                    let auction = match self.api.get_auction().await {
+                        Ok(auction) => auction,
+                        Err(_) => continue,
+                    };
+                    if auction.next_solver_competition != id {
+                        tracing::debug!("auction competition id has changed");
+                        break;
+                    }
+                }
+            };
+            if (tokio::time::timeout(Duration::from_secs(10), wait_for_id_to_change).await).is_err()
+            {
+                tracing::warn!(
+                    "auction competition id didn't change in time after submitting competition info"
+                );
+            }
+        }
```

Comment:
I think it would be nicer to move that code into a separate function.

Reply:
> The code seems to be correct but a little convoluted. Wouldn't it be simpler to store the last successfully solved `auction_id` in the driver and return early from every run loop where `current_auction_id <= last_solved_auction_id`?

Not sure. Then we have more "global" state (fields in the struct) and it becomes harder to handle the timeout because we'd also need to store "deadline after which we should no longer wait". To me it felt simpler to make this semantically part of the run that stores the competition info.

Reply:
Didn't mean to comment that in this thread. I will do the function refactor.